sql: decimal decode can panic if fed invalid byte sequences eg via pretty_key()

I'm not sure how to repro this error. But recording it here for posterity.

```
E190129 06:19:36.357429 4668183 sql/conn_executor.go:697  [n1,client=127.0.0.1:48924,user=root] a SQL panic has occurred while executing "SELECT crdb_internal.pretty_key(e'\\\\xa82a00918ed9':::BYTES, (-5096189069466142898):::INT8)": runtime error: index out of range
E190129 06:19:36.358168 4668183 util/log/crash_reporting.go:202  [n1,client=127.0.0.1:48924,user=root] a panic has occurred!
E190129 06:19:36.360857 4668183 util/log/crash_reporting.go:476  [n1,client=127.0.0.1:48924,user=root] Reported as error 014fbafa47f9403bb3bb323865572a1f
panic: runtime error: index out of range [recovered]
	panic: panic while executing 1 statements: SELECT crdb_internal.pretty_key(_:::BYTES, (_):::INT8); caused by runtime error: index out of range

goroutine 4668183 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc006fe6000, 0x30704e0, 0xc00c6a0600, 0x269d180, 0x46c09a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:711 +0x36d
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc006fe6000, 0x30704e0, 0xc00c6a0600)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:423 +0x61
panic(0x269d180, 0x46c09a0)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/util/encoding.makeDecimalFromMandE(0x0, 0x1, 0xc0024fddc2, 0x0, 0x6, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/encoding/decimal.go:374 +0x4b9
github.com/cockroachdb/cockroach/pkg/util/encoding.decodeDecimal(0xc0024fddc1, 0x5, 0x7, 0x0, 0x0, 0x0, 0x2a44500, 0x2, 0xc00c5eb848, 0x95744a, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/encoding/decimal.go:314 +0x5d3
github.com/cockroachdb/cockroach/pkg/util/encoding.DecodeDecimalAscending(0xc0024fddc1, 0x5, 0x7, 0x0, 0x0, 0x0, 0xc013894b80, 0x73, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/encoding/decimal.go:254 +0xb9
github.com/cockroachdb/cockroach/pkg/util/encoding.prettyPrintFirstValue(0x0, 0xc0024fddc1, 0x5, 0x7, 0xc00c5ebb08, 0x2, 0x2, 0x3, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/encoding/encoding.go:1484 +0x9ee
github.com/cockroachdb/cockroach/pkg/util/encoding.prettyPrintValueImpl(0x0, 0x0, 0x0, 0xc0024fddc0, 0x6, 0x8, 0x299856e, 0x1, 0xc01943ca80, 0x70, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/encoding/encoding.go:1396 +0x112
github.com/cockroachdb/cockroach/pkg/util/encoding.PrettyPrintValue(0x0, 0x0, 0x0, 0xc0024fddc0, 0x6, 0x8, 0x299856e, 0x1, 0x0, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/encoding/encoding.go:1360 +0x9b
github.com/cockroachdb/cockroach/pkg/keys.decodeKeyPrint(0x0, 0x0, 0x0, 0xc0024fddc0, 0x6, 0x8, 0x1, 0xc008747ae0)
	/go/src/github.com/cockroachdb/cockroach/pkg/keys/printer.go:493 +0xf7
github.com/cockroachdb/cockroach/pkg/keys.prettyPrintInternal.func1(0xc0024fddc0, 0x6, 0x8, 0x46f1e45, 0x1, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/keys/printer.go:529 +0x7f9
github.com/cockroachdb/cockroach/pkg/keys.prettyPrintInternal(0x0, 0x0, 0x0, 0xc0024fddc0, 0x6, 0x8, 0x203001, 0xc008747de0, 0x100000003074d20)
	/go/src/github.com/cockroachdb/cockroach/pkg/keys/printer.go:562 +0x2bc
github.com/cockroachdb/cockroach/pkg/keys.PrettyPrint(0x0, 0x0, 0x0, 0xc0024fddc0, 0x6, 0x8, 0x284e3c0, 0x10)
	/go/src/github.com/cockroachdb/cockroach/pkg/keys/printer.go:607 +0x65
github.com/cockroachdb/cockroach/pkg/roachpb.Key.StringWithDirs(0xc0024fddc0, 0x6, 0x8, 0x0, 0x0, 0x0, 0x0, 0x8913f8, 0x2711ec0)
	/go/src/github.com/cockroachdb/cockroach/pkg/roachpb/data.go:209 +0x80
github.com/cockroachdb/cockroach/pkg/sql/sqlbase.PrettyKey(0x0, 0x0, 0x0, 0xc0024fddc0, 0x6, 0x8, 0xb946b30ac63df74e, 0x2, 0x2)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sqlbase/keys.go:114 +0x76
github.com/cockroachdb/cockroach/pkg/sql/sem/builtins.glob..func216(0xc006fe6518, 0xc0081f5f40, 0x2, 0x2, 0x2, 0x2, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/builtins.go:2935 +0x143
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*FuncExpr).Eval(0xc01943ca10, 0xc006fe6518, 0xc01943ca10, 0x3075301, 0xc01943ca10, 0xc01943ca10)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/sem/tree/eval.go:3583 +0xb8
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildScalar(0xc008747db0, 0xc007f4eec0, 0x308b700, 0xc008747ae0, 0x1, 0x1, 0x2819240, 0x14e4268d7ce8)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/scalar_builder.go:99 +0x184
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildValues(0xc008747db0, 0xc00331fd88, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc0063764e0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:266 +0x217
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).buildRelational(0xc008747db0, 0x309f420, 0xc00331fd88, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00c5eca58, 0x89317f)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/relational_builder.go:155 +0x13b5
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).build(0xc008747db0, 0x307d2a0, 0xc00331fd88, 0xc006fe6518, 0xc00c5eca88, 0x893a08, 0x50)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:62 +0xab
github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder.(*Builder).Build(0xc008747db0, 0xc008747db0, 0xc006376180, 0xc00ec4c140, 0x3070500)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/execbuilder/builder.go:50 +0x41
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makeOptimizerPlan(0xc006fe6450, 0x30705a0, 0xc006376180, 0x29c2332, 0x19, 0x10409578604, 0x154c0587)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:140 +0x197
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).makeExecPlan(0xc006fe6000, 0x30705a0, 0xc006376180, 0xc006fe6450, 0x6, 0x89317f)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:929 +0x11b
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc006fe6000, 0x30705a0, 0xc006376180, 0xc006fe6450, 0x14e425637190, 0xc00a023680, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:824 +0x148
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc006fe6000, 0x30705a0, 0xc006376180, 0x3075ae0, 0xc007f4ecc0, 0xc006b2c5af, 0x58, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:437 +0xdd9
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc006fe6000, 0x30705a0, 0xc006376180, 0x3075ae0, 0xc007f4ecc0, 0xc006b2c5af, 0x58, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:93 +0x36f
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc006fe6000, 0x30704e0, 0xc00c6a0600, 0xc00a3719b8, 0x5400, 0x15000, 0xc00a371a50, 0xc00c264090, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1172 +0x21e5
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc0009b84e0, 0x30704e0, 0xc00c6a0600, 0xc006fe6000, 0x5400, 0x15000, 0xc00a371a50, 0xc00c264090, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:425 +0xce
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func4(0xc0009b84e0, 0x30704e0, 0xc00c6a0600, 0xc006fe6000, 0x5400, 0x15000, 0xc00a371a50, 0xc00c264090, 0xc00c2640c0, 0xc00204b2f4)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:319 +0x81
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:318 +0x1033
```

Found with RSG. https://teamcity.cockroachdb.com/viewLog.html?buildId=1113856&buildTypeId=Cockroach_Nightlies_RandomSyntaxTests

--- a/pkg/sql/logictest/testdata/logic_test/crdb_internal
+++ b/pkg/sql/logictest/testdata/logic_test/crdb_internal
@@ -366,3 +366,11 @@ query T
 select crdb_internal.node_executable_version()
 ----
 2.1-4
+
+user root
+
+# Regression test for #34441
+query T
+SELECT crdb_internal.pretty_key(e'\\xa82a00918ed9':::BYTES, (-5096189069466142898):::INT8);
+----
+/Table/32/???/9/6/81

--- a/pkg/util/encoding/decimal.go
+++ b/pkg/util/encoding/decimal.go
@@ -355,17 +355,16 @@ func getDecimalLen(buf []byte) (int, error) {
 // makeDecimalFromMandE reconstructs the decimal from the mantissa M and
 // exponent E.
 func makeDecimalFromMandE(negative bool, e int, m []byte, tmp []byte) (apd.Decimal, error) {
+	if len(m) == 0 {
+		return apd.Decimal{}, errors.New("expected mantissa, got zero bytes")
+	}
 	// ±dddd.
 	b := tmp[:0]
 	if n := len(m)*2 + 1; cap(b) < n {
 		b = make([]byte, 0, n)
 	}
-	for i, v := range m {
-		t := int(v)
-		if i == len(m) {
-			t--
-		}
-		t /= 2
+	for _, v := range m {
+		t := int(v) / 2
 		if t < 0 || t > 99 {
 			return apd.Decimal{}, errors.Errorf("base-100 encoded digit %d out of range [0,99]", t)
 		}

